cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -561,40 +605,102 @@ packageListFromCache mkPkg hnd Cache{..} mode = accum mempty [] mempty cacheEntr
 -- Index cache data structure
 --
 
+-- | Read the 'Index' cache from the filesystem
+readIndexCache :: Index -> IO Cache
+readIndexCache index
+  | is01Index index = decodeFile (cacheFile index)
```

Comment:
So what happens when reading fails? ie when a future version bumps the magic header number.

We ought somewhere to catch format exceptions from reading the cache and re-build the cache from scratch. Probably cleaner as a separate function that calls `readIndexCache` and `updatePackageIndexCacheFile` or something.